pkg/kubectl/rolling_updater.go

```diff
@@ -353,51 +411,94 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
 		}
 	}
 
-	// +1, -1 on oldRc, newRc until newRc has desired number of replicas or oldRc has 0 replicas
-	for newRc.Spec.Replicas < desired && oldRc.Spec.Replicas != 0 {
-		newRc.Spec.Replicas += 1
-		oldRc.Spec.Replicas -= 1
-		fmt.Printf("At beginning of loop: %s replicas: %d, %s replicas: %d\n",
-			oldName, oldRc.Spec.Replicas,
-			newName, newRc.Spec.Replicas)
-		fmt.Fprintf(out, "Updating %s replicas: %d, %s replicas: %d\n",
-			oldName, oldRc.Spec.Replicas,
-			newName, newRc.Spec.Replicas)
+	// Compute the scale amount based on a percentage of the new desired count.
+	// A negative percentage indicates our scale direction should be down-first.
+	scaleAmount := 1
+	skipFirstUp := false
+	if config.UpdatePercent != nil {
+		scaleAmount = int(math.Ceil(float64(desired) * (math.Abs(float64(*config.UpdatePercent)) / 100)))
+		if *config.UpdatePercent < 0 {
+			skipFirstUp = true
+		}
+	}
+	// Helpful output about what we're about to do.
+	direction := "up"
+	if skipFirstUp {
+		direction = "down"
+	}
+	fmt.Fprintf(out, "Scaling up %s from %d to %d, scaling down %s from %d to 0 (scale %s first by %d each interval)\n",
+		newRc.Name, newRc.Spec.Replicas, desired, oldRc.Name, oldRc.Spec.Replicas, direction, scaleAmount)
 
-		newRc, err = r.scaleAndWait(newRc, retry, waitForReplicas)
-		if err != nil {
-			return err
+	// Set up a function for scaling up.
+	scaleUp := func(newRc, oldRc *api.ReplicationController) (*api.ReplicationController, error) {
+		// Nothing to do.
+		if newRc.Spec.Replicas == desired {
```

Comment:
I thought it looked a little cleaner than making increasingly stateless functions which took a ton of params. Nevertheless...

Refactored to extract functions.